### PR TITLE
Changes to bowling

### LINKED ISSUE #80: Changes to Bowling tests
We recently rewrote the test suite for Bowling. Since this track implements Bowling, please [take a look at the new canonical_data.json file](https://github.com/exercism/x-common/blob/9049dc7f748fd038b140c878c4eaa89d09b73199/exercises/bowling/canonical-data.json) and see if your track should update its implementation or tests.

The new test suite reorders tests so that students can get to green quickly. It also reduces the number of tests so that students can focus on solving the interesting edge cases.

More details on this change are available in [this issue](https://github.com/exercism/x-common/issues/409)

### PR DESCRIPTION
**Pull request by dvberkel**
Work in progress. Eventually should fix #80

### CODE REVIEW

**Comment by tmcgilchrist** [comment]
Looks pretty good to me, big +1 for changing the exception to a data type.

The `bowling.mli` is missing an import for `result` by the looks of the failing test.

**Comment by dvberkel** [comment]
@tmcgilchrist thanks, I will look into the import issue, alongside the incorporation of all the tests

File: exercises/bowling/bowling.mli [hunk 1]
```diff
@@ -9,4 +9,4 @@ of pins knocked down. The return value is the updated state of the game. *)
 val roll : int -> t -> t
 
 (** This is called at the end of a game to retrieve the final score. *)
-val score : t -> int
+val score : t -> (int, string) result
```
Comment (stevejb71) [inline exercises/bowling/bowling.mli]: The canonical data doesn't allow for an error message. I'd suggest using a ```string option``` here.

Comment (dvberkel) [inline exercises/bowling/bowling.mli]: There is a discussion on the interpretation of errors and how to handle them. I think it is up to the maintainers to find a suitable means of expressing the error situations.

Having said that, the choice of encoding errors as `-1`, is unfortunate in my eyes.

None the less, I am still in favour `result` instead of `option` because I feel that the score, if there is one, should be an `int`.

**Comment by bcc32** [comment]
I would suggest using `int Or_error.t = (int, Error.t) result` instead of `(int, string) result`.  `Error.t` is basically like a wrapper around strings but is better suited for error handling due to lots of convenience functions (such as getting backtraces).

**Comment by stevejb71** [comment]
We should probably update the canonical data to include the error message if we go for a Result.t return type.

**Comment by dvberkel** [comment]
@bcc32 @stevejb71, both are good suggestions. Maybe best to have a working solution before we adopt it.

I will see if I can find the time to implement the last stretch.

And it would be nice to use thee generator too

**Comment by stevejb71** [comment]
Hi @dvberkel - do you want me to finish off this one? I wrote the original exercise example so it may be easier for me.

**Comment by dvberkel** [comment]
@stevejb71 please do. I am a bit swamped in work.


Will try to contribute again soon.

**Comment by stevejb71** [comment]
Fixed by #179
